### PR TITLE
Add Vue tests

### PR DESCRIPTION
**Pull request by afontcu**
Hi! 👋 As discussed in #119, I'm submitting a draft PR with a proof of concept for Vue tests. I just added tests for `userEvent.click`, as firstly I'd like to discuss the result and see if it's worth it :)

Tests do the exact same thing as React's. The main difference is the callback prop pattern (which I changed in favor of emitted events, way more idiomatic).

There are several things I'd like to highlight (and debate):

~1. List of emitted events doesn't include `mouse*` events, because AFAIK `wasAnotherElementFocused` is always false (Added code comments). Should I find a way to mount components differently, so I can check that mouse events are being sent?~

2. Also, `change` events are triggered twice. (e.g. [here](https://github.com/testing-library/user-event/pull/129/files#diff-d6a3965b2177d19c49ab5daaf670314cR78)).

3. In VTL, all `fireEvent.*` are converted into async functions. Here, they are (obviously) not. Not sure about the consequences, but I'm a bit worried. Yet, tests work 🤷‍♂.

I wasn't able to configure JSX for Vue (as it clashed with React's). So I first wrote tests using render functions, but I believe tagged templates are easily read. I wish I could use JSX, but it's OK I guess!

All in all, I feel it looks quite good! 😄 

So, (2) needs fixing, but I feel those are the kind of differences we were discussing in the issue linked above. (3) worries me the most. I guess @dfcook has the answer 😇 .

Any thoughts?

---

oh, and btw: there's an ongoing error with js-dom and `Error: Not implemented: HTMLFormElement.prototype.submit` ([related](https://github.com/jsdom/jsdom/issues/1937)). It's just a console warning, and tests are still passing. I think my changes had nothing to do with it 🤔

### CODE REVIEW

File: __tests__/vue/click.js [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+              mouseUp: eventsHandler,
+              click: eventsHandler
+            }
+          });
+        }
+      });
+
+      userEvent.click(getByTestId("element"));
+
+      // baseElement is always <body>, si wasAnotherElementFocused is false
```
Comment (afontcu) [inline __tests__/vue/click.js]: Missing events here:

File: __tests__/vue/click.js [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+            change: eventsHandler
+          }
+        });
+      }
+    });
+
+    userEvent.click(getByTestId("element"));
+
+    expect(events).toEqual([
+      // "mouseover",
```
Comment (afontcu) [inline __tests__/vue/click.js]: here:

File: __tests__/vue/click.js [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
```
Comment (afontcu) [inline __tests__/vue/click.js]: here:

File: __tests__/vue/click.js [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            mouseUp: eventsHandler,
+            click: eventsHandler
+          }
+        });
+      }
+    });
+
+    userEvent.click(getByTestId("div"));
+    expect(events).toEqual([
+      // "mouseover",
```
Comment (afontcu) [inline __tests__/vue/click.js]: ...and here:

**Comment by calebeby** [comment]
@afontcu For (4), it looks like the functions are all being converted to async so that they don't resolve till the next tick of the event loop (using `await wait()`), probably so that vue can run the event handlers and rerender before the next assertion runs

**Comment by Gpx** [comment]
Hey @afontcu, thanks for writing this. It seems Vue and React think very differently about events. What do you say if we export two different `userEvent` objects? One for Vue and one for React.

Something like: `import @testing-library/user-event/react`.

Eventually, if this library ends up being included in the main packages it should make sense

File: __tests__/vue/click.js [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+      const events = [];
+      const eventsHandler = jest.fn(evt => events.push(evt.type));
+      const { getByTestId } = render({
+        render: function(h) {
+          return h(type, {
+            attrs: {
+              "data-testid": "element"
+            },
+            on: {
+              mouseOver: eventsHandler,
```
Comment (calebeby) [inline __tests__/vue/click.js]:
```suggestion
              mouseover: eventsHandler,
```

@afontcu If you lowercase the handler properties, then they fire correctly (applies to all the events in this file)

Comment (calebeby) [reply __tests__/vue/click.js]: @afontcu I created a PR to your branch that fixes this: https://github.com/afontcu/user-event/pull/1

Comment (afontcu) [reply __tests__/vue/click.js]: Oh wow, I didn't even think about that. Really nice catch! 🙌

**Comment by afontcu** [comment]
Hi everybody!

Thanks to @calebeby PR, now tests are passing while tracking for the right events. This alleviates the differences between Vue and React implementations, I think.

Do you folks think that exporting two different objects is the best approach here? I guess it'd work, but it'd be amazing if we could find the way for a single object to work.

What's missing from my point of view is that double `change` events being emitted. Apart from that, are we on the right track to provide a single object for both Vue and React?



Thanks all for your inputs (no pun intended)! 😃

**Comment by calebeby** [comment]
I don't think that it is necessary for `user-event` to have separate versions for different libraries. If `user-event` fires the same events that real users do, then it should be compatible with _all_ libraries.

I created a couple codesandbox's that we could use to check our work against the real events that browsers fire:

https://codesandbox.io/s/input-typecheckbox-events-9m8ts
https://codesandbox.io/s/input-typetext-events-tkftr

**Comment by afontcu** [comment]
I'm closing this since a new, rebased PR has been merged. I'll open up a new one with additional tests. Thanks all for the feedback!

**Pull request closed without merging**